Added full end to end test for traffic analytics page hit tracking

cmraible opened the pull request:
ref https://linear.app/ghost/issue/PROD-2687/

This PR adds the first full end to end test for event tracking:
- Visit site's homepage
- Visit Admin > Analytics
- Check that a unique visitor is recorded

It adds two services to the compose file:
- analytics service, to proxy events to Tinybird
- caddy, to route requests from /.ghost/analytics/* to the analytics service

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds an end-to-end test that records a homepage visit and verifies a unique visitor via a Caddy-proxied analytics endpoint, plus Docker/Tinybird/CI and helper refactors to support it.
> 
> - **E2E/Testing**:
>   - Add test to record a homepage visit and assert 1 unique visitor in `admin/analytics`.
>   - New helpers: `withIsolatedPage`, `PublicPage.enableAnalyticsRequests()` and `waitForPageHitRequest()`; enhanced `AnalyticsOverviewPage` with `UniqueVisitorsGraph` and `refreshData()`.
>   - Add `data-testid` hooks in `apps/shade` KPI card (`kpi-card-header-value`, `kpi-card-header-diff`).
> - **Docker/Compose**:
>   - Add `e2e/compose.yml` with `caddy` and `analytics` services; route `/.ghost/analytics/*` to analytics via new `.docker/caddy/Caddyfile.e2e`.
>   - Switch analytics healthchecks to Node fetch; expose `tinybird-local` port.
> - **E2E Infra**:
>   - Refactor `DockerCompose`: project scoping, `up()`, `waitForAll()`, status parsing, improved logs; update `down()`/`run` commands to include `-p`.
>   - `EnvironmentManager`: use `up()` + `waitForAll()`; clarify teardown.
>   - `GhostManager`: set `tinybird__tracker__endpoint` to `http://localhost:8080/.ghost/analytics/api/v1/page_hit` and `tinybird__tracker__datasource`.
> - **Tinybird CLI**:
>   - Harden token retrieval: single request, error checks, select admin token by ADMIN scope, reuse response.
> - **CI**:
>   - Use `ubuntu-latest` runners; update compose path to `e2e/compose.yml` for image pulls.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 1cf1f2412ce0fc21fba625b6939ea7305c3b4849. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

cmraible committed:
Added analytics service to e2e tests and fixed healthcheck

cmraible committed:
Simplified implementation of waitForAll method

cmraible committed:
Added first full end to end page hit tracking test

cmraible committed:
Fixed linter errors

cmraible committed:
Added explanation for why docker compose up --wait isn't used

cmraible committed:
Cleaned up and improved comments in environment manager

cmraible committed:
Cleaned up page hit tracking test and moved into overview tab test

ibalosh reviewed (changes requested):
Hey @cmraible , 

In the review I mostly focused on the readability of the tests, and some basic cleanups we could do in the environment helper files. I did not go too deep there, since it's still new and we are trying it out - thought I would clean them up a bit in terms of extracting some of the methods and use better names for variable names.

Overall I like the idea (sad test containers did not work - will love to hear more). 

My main requests for changes is for decoupling more implementation details from tests, to make them less aware of internals (in terms of logic, for refreshing etc). We should try to hide as much as we can from the test itself.

For some reason I could not get the test to pass, I kept getting times. Let's catchup when you are back.

ibalosh commented on e2e/tests/admin/analytics/page-hit-tracking.test.ts:
```diff
@@ -0,0 +1,43 @@
+        // Set the synthetic monitoring flag to bypass test environment detection
+        await publicPage.addInitScript(() => {
+            window.__GHOST_SYNTHETIC_MONITORING__ = true;
+        });
```
yeah saw that, it's fine as long as we describe the intention, like in `browser-service.js` , where we say `isTestEnvironment`. Shared details how you can make it clearer below in comments.

cmraible committed:
Renamed e2e/compose.e2e.yml to e2e/compose.yml

cmraible committed:
Added withIsolatedPage helper

cmraible committed:
Added enableAnalyticsRequests() to base PublicPage

cmraible committed:
Added waitForPageHitRequest to PublicPage

cmraible committed:
Added refreshData method to AnalyticsOverviewPage

cmraible committed:
Renamed `cmd` to `command`

cmraible committed:
Refactored waitForAll method

cmraible commented on e2e/tests/admin/analytics/overview.test.ts:
```diff
@@ -4,8 +4,40 @@ import {
+        // Ghost does not send tracking requests if it detects that it is running in a test environment, i.e. playwright
+        // We can override this by setting window.__GHOST_SYNTHETIC_MONITORING__ to true
+        // This allows us to test page hit tracking in our e2e tests
+        const publicContext = await browser.newContext({
```
Awesome suggestions! I've made a few changes here:

- created and used `withIsolatedPage()` helper, agree that this is much cleaner
- rather than creating a separate helper for `bypassTestEnvironmentSettingsForPage()`, I've just added this to the constructor of the `PublicPage` POM, which should be an ancestor to all pages that need analytics events to be enabled.

cmraible commented on e2e/tests/admin/analytics/overview.test.ts:
```diff
@@ -4,8 +4,40 @@ import {
+        const homePage = new HomePage(publicPage);
+        await homePage.goto();
+        await expect(homePage.title).toBeVisible();
+        await publicPage.waitForLoadState('networkidle');
```
Good shout - what we really need to wait for is the POST request to `/.ghost/analytics/api/v1/page_hit`. I've reworked this a bit by adding a `waitForPageHitRequest()` method to the `PublicPage` POM, and used it here, thus eliminating the title and load state checks here.

ibalosh replied:
gotcha, tx!

cmraible commented on e2e/tests/admin/analytics/overview.test.ts:
```diff
@@ -4,8 +4,40 @@ import {
+        // Reload to ensure we fetch the latest data from Tinybird
+        const analyticsOverviewPage = new AnalyticsOverviewPage(page);
+        await analyticsOverviewPage.goto();
+        await page.reload();
```
Agreed! Added a `refreshData()` method to the POM and used it here.

cmraible commented on e2e/helpers/environment/DockerCompose.ts:
```diff
@@ -17,30 +17,96 @@ export class DockerCompose {
+            const output = execSync(cmd, {encoding: 'utf-8'}).trim();
+
+            if (!output) {
+                await new Promise((resolve) => {
```
I've made a few improvements to this area to reduce the duplication and make it (hopefully) read a lot clearer 👌

cmraible committed:
Fixed path to compose file in CI

cmraible committed:
Fixed call to async method in constructor of PublicPage

ibalosh reviewed (approved):
that looks great! 👍

cmraible committed:
Updated GHA runner to use default instead of ubuntu-latest-16-cores

The average queue time for the ubuntu-latest-16-cores runners is almost 2 minutes, which completely offsets any benefit from using the more powerful runners

cmraible committed:
Added better error logging when docker compose up fails

cmraible committed:
Improved error logging again

cmraible committed:
Added full docker compose logs for debugging on failure

cmraible committed:
Added debug logs to tb-cli entrypoint

cmraible committed:
Fixed fragile token fetching in tb entrypoint script

cmraible committed:
Cleaned up tb-cli entrypoint script
